Security Fix: set begin flag before calling provider

quentinus95 opened the pull request:
Hello @scheb 

I was experiencing a security issue in my Symfony app, this patch fixes it.

When calling `beginAuthentication` method on a custom provider, my application was throwing an exception because of a service inavailability my custom provider depends on (a cURL call was failing). Because the flag is set only following the method return value, no flag was set at this point.

The consequence is that the user was able to change the url and go to the application without completing double authentication. To avoid the error, I set the flag before calling the method so if the call never returns, the flag is set and the user cannot have access to the application.

quentinus95 commented:
If you are OK with the change, I will fix the single failing test.

scheb commented:
In general looks good to me.

I do not like that you're using setComplete(), which sets a session state, as if the user would have passed the 2fa step, although it should never have been started. I'd prefer a new method on the session flag manager, to unset the flag in that use-case, instead of setting it true.

quentinus95 commented:
@scheb I've just updated my code. I've tested it in my app, working as expected. I have fixed and added a new test.

The failing build is not linked to the change.

scheb commented:
Thanks for your contribution!
